lan865x/src/esp_eth_mac_lan865x.c

```diff
@@ -0,0 +1,1149 @@
+                data++;
+            }
+        }
+err:
+        lan865x_spi_unlock(emac);
+    } else {
+        ESP_LOGE(TAG, "Failed to lock SPI");
+        ret = ESP_ERR_TIMEOUT;
+    }
+    return ret;
+}
+
+static esp_err_t lan865x_read_reg(emac_lan865x_t *emac, uint8_t mms, uint16_t addr, uint32_t *value)
+{
+    return lan865x_control_transaction(emac, LAN865X_READ_REG, mms, addr, 1, value);
+}
+
+static esp_err_t lan865x_write_reg(emac_lan865x_t *emac, uint8_t mms, uint16_t addr, uint32_t value)
+{
+    return lan865x_control_transaction(emac, LAN865X_WRITE_REG, mms, addr, 1, &value);
+}
+
+static esp_err_t lan865x_set_reg_bits(emac_lan865x_t *emac, uint8_t mms, uint16_t addr, uint32_t bitmask)
+{
+    esp_err_t ret = ESP_OK;
+    uint32_t reg_value;
+
+    ESP_GOTO_ON_ERROR(lan865x_read_reg(emac, mms, addr, &reg_value), err, TAG, "Failed to read register MMS: %" PRIu8 ", ADDR: 0x%04" PRIx16, mms, addr);
+    reg_value |= bitmask;
+    ESP_GOTO_ON_ERROR(lan865x_write_reg(emac, mms, addr, reg_value), err, TAG, "Failed to write register MMS: %" PRIu8 ", ADDR: 0x%04" PRIx16, mms, addr);
+err:
+    return ret;
+}
+
+static esp_err_t lan865x_clear_reg_bits(emac_lan865x_t *emac, uint8_t mms, uint16_t addr, uint32_t bitmask)
+{
+    esp_err_t ret = ESP_OK;
+    uint32_t reg_value;
+    ESP_GOTO_ON_ERROR(lan865x_read_reg(emac, mms, addr, &reg_value), err, TAG, "Failed to read register MMS: %" PRIu8 ", ADDR: 0x%04" PRIx16, mms, addr);
+    reg_value &= ~bitmask;
+    ESP_GOTO_ON_ERROR(lan865x_write_reg(emac, mms, addr, reg_value), err, TAG, "Failed to write register MMS: %" PRIu8 ", ADDR: 0x%04" PRIx16, mms, addr);
+err:
+    return ret;
+}
+
+// !! Proprietary access mechanism. Do not confuse this with the Clause 22 indirect access to Clause 45 registers. !!
+static esp_err_t lan865x_indirect_read(emac_lan865x_t *emac, uint8_t addr, uint8_t mask, uint8_t *value)
+{
+    esp_err_t ret = ESP_OK;
+    uint32_t reg_value;
+    ESP_GOTO_ON_ERROR(lan865x_write_reg(emac, 0x04, 0x00D8, addr), err, TAG, "Failed to write register MMS: 0x04, ADDR: 0x00D8");
+    ESP_GOTO_ON_ERROR(lan865x_write_reg(emac, 0x04, 0x00DA, 0x2), err, TAG, "Failed to write register MMS: 0x04, ADDR: 0x00DA");
+    ESP_GOTO_ON_ERROR(lan865x_read_reg(emac, 0x04, 0x00D9, &reg_value), err, TAG, "Failed to read register MMS: 0x04, ADDR: 0x00D9");
+    *value = reg_value & mask;
+err:
+    return ret;
+}
+
+// Configuration recommended by manufacturer (http://www.microchip.com/DS60001760)
+static esp_err_t lan865x_default_config(emac_lan865x_t *emac)
+{
+    esp_err_t ret = ESP_OK;
+
+    // Read configuration parameters
+    uint8_t value1, value2;
+    int8_t offset1, offset2;
+    uint16_t cfgparam1, cfgparam2;
+
+    // Read value1 from address 0x04
+    ESP_GOTO_ON_ERROR(lan865x_indirect_read(emac, 0x04, 0x1F, &value1), err, TAG, "Failed to read value1");
+
+    // Calculate offset1
+    if ((value1 & 0x10) != 0) {
+        offset1 = (int8_t)((uint8_t)value1 - 0x20);
```

Comment:
Question: Why casting `value1` to unsigned? the other operand has no (0x20U), so it would be still signed subtraction, correct?